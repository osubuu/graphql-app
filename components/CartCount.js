import PropTypes from 'prop-types';
import styled from 'styled-components';
import { TransitionGroup, CSSTransition } from 'react-transition-group';

const Dot = styled.div`
  background: ${props => props.theme.red};
  color: white;
  border-radius: 50%;
  padding: 0.5rem;
  line-height: 2rem;
  min-width: 3rem;
  margin-left: 1rem;
  font-weight: 100;

  /* keep character size of each number constant
  (e.g. skinny '1' will keep same width as larger '2) */
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;
`;

const AnimationStyles = styled.span`
  position: relative;
  .count {
    display: block;
    position: relative;
    transition: all 0.4s;
    backface-visibility: hidden;
  }
  /* Initial state of the entered Dot (flip on its back) */
  .count-enter {
    /* 0.5turn = 180 degrees */
    transform: scale(4) rotateX(0.5turn);
  }
  .count-enter-active {
    transform: rotateX(0);
  }
  .count-exit {
    position: absolute;
    top: 0;
    transform: rotateX(0);
  }
  .count-exit-active {
    transform: scale(4) rotateX(0.5turn);
  }
`;

const CartCount = ({ count }) => (
  <AnimationStyles>
    <TransitionGroup>
      <CSSTransition
        unmountOnExit
        className="count"
        classNames="count"
        key={count}
        timeout={{ enter: 400, exit: 400 }}
      >
        <Dot>{count}</Dot>
      </CSSTransition>
    </TransitionGroup>
  </AnimationStyles>
);

CartCount.propTypes = {
  count: PropTypes.number.isRequired,
};

export default CartCount;
