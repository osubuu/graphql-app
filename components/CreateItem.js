import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Router from 'next/router';

import Form from './styles/Form';
import ErrorMessage from './ErrorMessage';
import { PAGINATION_QUERY } from './Pagination';

// recall: these mutations and queries were defined in schema.grapql yoga file in the back-end
export const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`;

class CreateItem extends Component {
  state = {
    title: '',
    description: '',
    image: '',
    largeImage: '',
    price: 0,
  }

  handleChange = e => {
    const { name, type, value } = e.target;
    const val = type === 'number' && value ? parseFloat(value) : value;
    this.setState({ [name]: val });
  }

  uploadFile = async e => {
    const { files } = e.target;
    const data = new FormData();
    data.append('file', files[0]);
    data.append('upload_preset', 'graphql-app');

    const res = await fetch('https://api.cloudinary.com/v1_1/df7nqbjav/image/upload/', {
      method: 'POST',
      body: data,
    });

    const file = await res.json();
    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url,
    });
  }

  render() {
    return (
      <Mutation
        mutation={CREATE_ITEM_MUTATION}
        refetchQueries={[
          { query: PAGINATION_QUERY },
        ]}
        variables={this.state}
      >
        {(createItem, { loading, error }) => (
          <Form
            data-test="form"
            onSubmit={async e => {
              e.preventDefault();
              const res = await createItem(); // sends the mutation request
              Router.push({
                pathname: '/item',
                query: { id: res.data.createItem.id },
              });
            }}
          >
            <ErrorMessage error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="file">
                Image
                <input
                  type="file"
                  id="file"
                  name="file"
                  required
                  placeholder="Upload an image"
                  onChange={this.uploadFile}
                />
                {this.state.image && (
                <img
                  width="200"
                  src={this.state.image}
                  alt="Upload Preview"
                />
                )}
              </label>

              <label htmlFor="title">
                Title
                <input
                  type="text"
                  id="title"
                  name="title"
                  required
                  placeholder="Title"
                  value={this.state.title}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="price">
                Price
                <input
                  type="number"
                  id="price"
                  name="price"
                  required
                  placeholder="Price"
                  value={this.state.price}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="description">
                Description
                <input
                  id="description"
                  name="description"
                  required
                  placeholder="Enter a description"
                  value={this.state.description}
                  onChange={this.handleChange}
                />
              </label>
              <button type="submit">Submit</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default CreateItem;
