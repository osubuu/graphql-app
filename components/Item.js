import PropTypes from 'prop-types';
import Link from 'next/link';

import Title from './styles/Title';
import ItemStyles from './styles/ItemStyles';
import PriceTag from './styles/PriceTag';
import formatMoney from '../lib/formatMoney';

import DeleteItem from './DeleteItem';
import AddToCart from './AddToCart';

const Item = props => {
  const { item, me } = props;
  const canEditAndDelete = me && me.id === item.user.id;
  return (
    <ItemStyles>
      <Link href={{
        pathname: '/item',
        query: { id: item.id },
      }}
      >
        <a>
          {item.image && <img src={item.image} alt={item.title} />}
        </a>
      </Link>
      <Title>
        <Link href={{
          pathname: '/item',
          query: { id: item.id },
        }}
        >
          <a>{item.title}</a>
        </Link>
      </Title>
      <PriceTag>{formatMoney(item.price)}</PriceTag>
      <p>{item.description}</p>

      <div className="buttonList">
        {canEditAndDelete && (
        <Link href={{
          pathname: 'update',
          query: { id: item.id },
        }}
        >
          <a>Edit</a>
        </Link>
        )}
        <AddToCart id={item.id} />
        {canEditAndDelete && <DeleteItem id={item.id}>Delete This Item</DeleteItem>}
      </div>
    </ItemStyles>
  );
};

Item.propTypes = {
  item: PropTypes.oneOfType([PropTypes.object]).isRequired,
};

export default Item;
