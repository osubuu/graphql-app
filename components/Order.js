import PropTypes from 'prop-types';
import { Query } from 'react-apollo';
import { format } from 'date-fns';
import Head from 'next/head';
import gql from 'graphql-tag';
import formatMoney from '../lib/formatMoney';
import ErrorMessage from './ErrorMessage';
import OrderStyles from './styles/OrderStyles';

export const SINGLE_ORDER_QUERY = gql`
  query SINGLE_ORDER_QUERY($id: ID!) {
    order(id: $id) {
      id
      charge
      total
      createdAt
      user {
        id
      }
      items {
        id
        title
        description
        price
        image
        quantity
      }
    }
  }
`;

const Order = props => (
  <Query query={SINGLE_ORDER_QUERY} variables={{ id: props.id }}>
    {({ data, error, loading }) => {
      if (error) return <ErrorMessage error={error} />;
      if (loading) return <p>Loading...</p>;
      const { order } = data;
      return (
        <OrderStyles data-test="order">
          <Head>
            <title>Sick Fits - Order {order.id}</title>
          </Head>
          <p>
            <span>Order ID:</span>
            <span>{order.id}</span>
          </p>
          <p>
            <span>Charge</span>
            <span>{order.charge}</span>
          </p>
          <p>
            <span>Date</span>
            <span>{format(order.createdAt, 'MMMM d, YYYY h:mm a')}</span>
          </p>
          <p>
            <span>Order Total</span>
            <span>{formatMoney(order.total)}</span>
          </p>
          <p>
            <span>Item Count</span>
            <span>{order.items.length}</span>
          </p>
          <div className="items">
            {order.items.map(item => (
              // TODO: make it its own component
              <div className="order-item" key={item.id}>
                <img src={item.image} alt={item.title} />
                <div className="item-details">
                  <h2>{item.title}</h2>
                  <p>Qty: {item.quantity}</p>
                  <p>Each: {formatMoney(item.price)}</p>
                  <p>Subtotal: {formatMoney(item.price * item.quantity)}</p>
                  <p>{item.description}</p>
                </div>
              </div>
            ))}
          </div>
        </OrderStyles>
      );
    }}
  </Query>
);

Order.propTypes = {
  id: PropTypes.string.isRequired,
};

export default Order;
