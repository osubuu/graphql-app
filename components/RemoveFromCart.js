import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import { CURRENT_USER_QUERY } from './User';

export const REMOVE_FROM_CART_MUTATION = gql`
  mutation removeFromCart($id: ID!) {
    removeFromCart(id: $id) {
      id
    }
  }
`;

const BigButton = styled.button`
  font-size: 3rem;
  background: none;
  border: 0;
  &:hover {
    color: ${props => props.theme.red};
    cursor: pointer;
  }
`;

class RemoveFromCart extends Component {
  // this gets called as soon as we get a response back from the server
  // after a mutation has been performed
  update = (cache, payload) => {
    // 1. First read the cache
    const data = cache.readQuery({ query: CURRENT_USER_QUERY });
    // 2. Remove that item from the cart
    const cartItemId = payload.data.removeFromCart.id;
    data.me.cart = data.me.cart.filter(cartItem => cartItem.id !== cartItemId);
    // 3. Write it back to the cache
    cache.writeQuery({ query: CURRENT_USER_QUERY, data });
  }

  render() {
    return (
      <Mutation
        mutation={REMOVE_FROM_CART_MUTATION}
        variables={{ id: this.props.id }}
        update={this.update}
        // run a function preemptively ASSUMING the mutation would be a success
        // this is for optimal UX
        optimisticResponse={{
          __typename: 'Mutation',
          removeFromCart: {
            __typename: 'CartItem',
            id: this.props.id,
          },
        }}
      >
        {(removeFromCart, { loading }) => (
          <BigButton
            title="Delete Item"
            disabled={loading}
            onClick={() => {
              removeFromCart().catch(err => alert(err.message));
            }}
          >
            &times;
          </BigButton>
        )}
      </Mutation>
    );
  }
}

RemoveFromCart.propTypes = {
  id: PropTypes.string.isRequired,
};

export default RemoveFromCart;
