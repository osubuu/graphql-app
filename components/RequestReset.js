import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';

import Form from './styles/Form';
import ErrorMessage from './ErrorMessage';
import { Fieldset } from './Signin';

export const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($email: String!) {
    requestReset(email: $email) {
      message
    }
  }
`;

class RequestReset extends Component {
  state = {
    email: '',
  }

  saveToState = e => {
    this.setState({ [e.target.name]: e.target.value });
  }

  render() {
    return (
      <Mutation
        mutation={REQUEST_RESET_MUTATION}
        variables={this.state}
      >
        {(reset, { error, loading, called }) => (
          <Form
            data-test="form"
            method="post"
            onSubmit={async e => {
              e.preventDefault();
              await reset();
              this.setState({ email: '' });
            }}
          >
            <Fieldset disabled={loading} aria-busy={loading}>
              <div>
                <h2>Request a password reset</h2>
                <ErrorMessage error={error} />
                {!error && !loading && called && <p>Success! Check your email for a reset link!</p>}
                <label htmlFor="email">
                  Email
                  <input
                    type="email"
                    name="email"
                    placeholder="email"
                    value={this.state.email}
                    onChange={this.saveToState}
                  />
                </label>
                <button type="submit">Request reset</button>
              </div>
            </Fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default RequestReset;
