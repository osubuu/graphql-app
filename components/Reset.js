import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';

import Form from './styles/Form';
import ErrorMessage from './ErrorMessage';
import { CURRENT_USER_QUERY } from './User';

const RESET_MUTATION = gql`
  mutation RESET_MUTATION($resetToken: String!, $oldPassword: String!, $password: String!, $confirmPassword: String!) {
    resetPassword(resetToken: $resetToken, oldPassword: $oldPassword, password: $password, confirmPassword: $confirmPassword) {
      id
      email
      name
    }
  }
`;

class Reset extends Component {
  state = {
    oldPassword: '',
    password: '',
    confirmPassword: '',
  }

  saveToState = e => {
    this.setState({ [e.target.name]: e.target.value });
  }

  render() {
    return (
      <Mutation
        mutation={RESET_MUTATION}
        variables={{
          resetToken: this.props.resetToken,
          oldPassword: this.state.oldPassword,
          password: this.state.password,
          confirmPassword: this.state.confirmPassword,
        }}
        refetchQueries={[
          { query: CURRENT_USER_QUERY },
        ]}
      >
        {(reset, { error, loading, called }) => {
          const passwordReset = !error && !loading && called;
          return (
            <Form
              method="post"
              onSubmit={async e => {
                e.preventDefault();
                await reset();
                this.setState({
                  oldPassword: '',
                  password: '',
                  confirmPassword: '',
                });
              }}
            >
              <fieldset disabled={loading} aria-busy={loading}>
                <h2>Reset your password</h2>
                <ErrorMessage error={error} />
                {!passwordReset ? (
                  <>
                    <label htmlFor="oldPassword">
                      Current password
                      <input
                        type="password"
                        name="oldPassword"
                        placeholder="current password"
                        value={this.state.oldPassword}
                        onChange={this.saveToState}
                      />
                    </label>
                    <label htmlFor="password">
                      New password
                      <input
                        type="password"
                        name="password"
                        placeholder="new password"
                        value={this.state.password}
                        onChange={this.saveToState}
                      />
                    </label>
                    <label htmlFor="confirmPassword">
                      Confirm your new password
                      <input
                        type="password"
                        name="confirmPassword"
                        placeholder="confirm new password"
                        value={this.state.confirmPassword}
                        onChange={this.saveToState}
                      />
                    </label>
                    <button type="submit">Reset</button>
                  </>
                ) : <p>Password reset successfully!</p> }
              </fieldset>
            </Form>
          );
        }}
      </Mutation>
    );
  }
}

Reset.propTypes = {
  resetToken: PropTypes.string.isRequired,
};

export default Reset;
