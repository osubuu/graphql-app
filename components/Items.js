import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import styled from 'styled-components';
import { adopt } from 'react-adopt';

import Pagination from './Pagination';
import User from './User';
import Item from './Item';
import { perPage } from '../config';

// best practice to put queries in CAPS
export const ALL_ITEMS_QUERY = gql`
  query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = ${perPage}) {
    items(first: $first, skip: $skip, orderBy: createdAt_DESC) {
      id
      title
      price
      description
      image
      largeImage
      user {
        id
      }
    }
  }
`;

const Center = styled.div`
  text-align: center;
`;

const ItemsList = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 60px;
  max-width: ${props => props.theme.maxWidth};
  margin: 0 auto;
`;


const Items = props => {
  const Composed = adopt({
    user: ({ render }) => <User>{render}</User>,
    items: ({ render }) => (
      <Query
        query={ALL_ITEMS_QUERY}
        variables={{
          skip: props.page * perPage - perPage,
        }}
      >{render}
      </Query>
    ),
  });

  return (
    <Composed>
      {({ user, items }) => {
        if (items.loading) return <p>Loading...</p>;
        if (items.error) return <p>Error: {items.error.message}</p>;
        return (
          <Center>
            <Pagination page={props.page} />
            <ItemsList>
              {items.data.items.map(item => <Item item={item} key={item.id} me={user.data.me} />)}
            </ItemsList>
            <Pagination page={props.page} />
          </Center>
        );
      }}
    </Composed>
  );
};

export default Items;
