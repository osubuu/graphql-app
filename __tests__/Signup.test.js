import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import { ApolloConsumer } from 'react-apollo';
import Router from 'next/router';

import Signup, { SIGNUP_MUTATION } from '../components/Signup';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser } from '../lib/testUtils';

function type(wrapper, name, value) {
  wrapper.find(`input[name="${name}"]`).simulate('change', {
    target: { name, value },
  });
}

const me = fakeUser();
const mocks = [
  // signup mutation mock
  {
    request: {
      query: SIGNUP_MUTATION,
      variables: {
        email: me.email,
        name: me.name,
        password: 'password1234',
      },
    },
    result: {
      data: {
        signup: {
          __typename: 'User',
          id: 'abc123',
          email: me.email,
          name: me.name,
        },
      },
    },
  },
  // current user query mock
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me } },
  },
];

describe('<Signup/>', () => {
  it('renders and matches snapshot', async () => {
    const wrapper = mount(
      <MockedProvider>
        <Signup />
      </MockedProvider>,
    );
    expect(toJSON(wrapper.find('form'))).toMatchSnapshot();
  });

  it('calls the mutation properly', async () => {
    let apolloClient;
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <ApolloConsumer>
          {client => {
            apolloClient = client;
            return <Signup />;
          }}
        </ApolloConsumer>
      </MockedProvider>,
    );
    await wait();
    wrapper.update();
    type(wrapper, 'name', me.name);
    type(wrapper, 'email', me.email);
    type(wrapper, 'password', 'password1234');
    wrapper.update();
    Router.router = { push: jest.fn() };
    wrapper.find('form').simulate('submit');
    await wait(); // recall: this is used to prevent weird race conditions
    // query the user out of the apollo client
    const user = await apolloClient.query({ query: CURRENT_USER_QUERY });
    expect(user.data.me).toMatchObject(me);
    expect(Router.router.push).toHaveBeenCalledWith({
      pathname: '/',
    });
  });
});
