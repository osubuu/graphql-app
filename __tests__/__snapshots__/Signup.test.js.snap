// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup/> renders and matches snapshot 1`] = `
<form
  className="Form-sc-1ovjk64-0 cEPOGN"
  method="post"
  onSubmit={[Function]}
>
  <Signin__Fieldset
    aria-busy={false}
    disabled={false}
  >
    <fieldset
      aria-busy={false}
      className="Signin__Fieldset-sc-1bn107d-0 eFAYSk"
      disabled={false}
    >
      <div>
        <h2>
          Sign up for an Account
        </h2>
        <DisplayError
          error={Object {}}
        />
        <label
          htmlFor="email"
        >
          Email
          <input
            name="email"
            onChange={[Function]}
            placeholder="email"
            type="email"
            value=""
          />
        </label>
        <label
          htmlFor="name"
        >
          Name
          <input
            name="name"
            onChange={[Function]}
            placeholder="name"
            type="text"
            value=""
          />
        </label>
        <label
          htmlFor="password"
        >
          Password
          <input
            name="password"
            onChange={[Function]}
            placeholder="password"
            type="password"
            value=""
          />
        </label>
        <button
          type="submit"
        >
          Sign up
        </button>
      </div>
    </fieldset>
  </Signin__Fieldset>
</form>
`;
